### PR TITLE
Reland (x2) "Output .js files as ES6 modules. (flutter#52023)"

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
Second attempt to reland https://github.com/flutter/engine/pull/52023

Fixes since the previous reland attempt:
* We need to pass the skwasm main JS URI when loading the module so that it can pass that along to the worker. Since the worker uses the workaround to allow a cross script worker, it has trouble locating the main JS URI in relation to itself in a way that actually works for dynamic imports, so passing it along fixes that issue.
* Some of the Google3 tests relied on the relative default canvaskit path. Dynamic module imports seems to not handle relative paths the way we expect, so we do our own URL resolution using the URL constructor before passing it into the dynamic import API. Also cleaned up some of the other relative pathing stuff that we do around the base URI. in flutter.js

### CODE REVIEW

**Comment by kevmoo** [comment]
🤞

**Review by ditman — APPROVED** [review]
LGTM!

File: lib/web_ui/flutter_js/src/utils.js [hunk 1]
```diff
@@ -54,5 +51,5 @@ export function getCanvaskitBaseUrl(config, buildConfig) {
   if (buildConfig.engineRevision && !buildConfig.useLocalCanvasKit) {
     return joinPathSegments("https://www.gstatic.com/flutter-canvaskit", buildConfig.engineRevision);
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/utils.js:52]: Ah yeah, I agree. Although I'm not going to tackle that with this change (and would require other changes in the framework first).

[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/utils.js:52]: 100% agreed

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 2]
```diff
@@ -3698,7 +3698,7 @@ Future<CanvasKitModule> _downloadOneOf(Iterable<String> urls) async {
 }
 
 String _resolveUrl(String url) {
-  return DomURL(url.toJS, domWindow.document.baseUri?.toJS).toJSString().toDart;
+  return createDomURL(url, domWindow.document.baseUri).toJSString().toDart;
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:3701]: You could wrap the `.toJSString().toDart` in a `toString()` method in the `DomURLExtension` :)

**Pull request merged**